src/relay/ir/text_printer.cc

```diff
@@ -0,0 +1,752 @@
+   *
+   * This is to avoid the call of subsequent GetValue print out
+   * additional instructions which get mixed with the partial instruction
+   * printed by the caller.
+   *
+   * \param expr The input expression.
+   * \return The text value of Expr.
+   */
+  TextValue GetValue(const Expr& expr) {
+    auto it = memo_.find(expr);
+    if (it != memo_.end()) return it->second;
+    TextValue val = this->VisitExpr(expr);
+    memo_[expr] = val;
+    return val;
+  }
+  //------------------------------------
+  // Overload of Expr printing functions
+  //------------------------------------
+  TextValue VisitExpr_(const ConstantNode* op) final {
+    // Print out simple scalar directly.
+    if (op->is_scalar()) {
+      std::ostringstream os;
+      DataType dtype = TVMType2Type(op->data->dtype);
+      CHECK_EQ(op->data->ctx.device_type, kDLCPU);
+      if (dtype == Int(32)) {
+        return ConstScalar(dtype, static_cast<const int32_t*>(op->data->data));
+      } else if (dtype == Int(64)) {
+        return ConstScalar(dtype, static_cast<const int64_t*>(op->data->data));
+      } else if (dtype == Float(32)) {
+        return ConstScalar(dtype, static_cast<const float*>(op->data->data));
+      } else if (dtype == Float(64)) {
+        return ConstScalar(dtype, static_cast<const double*>(op->data->data));
+      }
+    }
+    // default fall-back, record it as meta node.
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = " << meta_.GetMetaNode(GetRef<NodeRef>(op));
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const TupleNode* op) final {
+    std::vector<TextValue> fields;
+    for (Expr field : op->fields) {
+      fields.push_back(GetValue(field));
+    }
+    // NOTE: always recursively visit to get ids,
+    // before print out the current line
+    TextValue id = this->AllocTempVar();
+    this->PrintIndent();
+    stream_ << id << " = (";
+    for (size_t i = 0; i < fields.size(); ++i) {
+      stream_ << fields[i];
+      if (i + 1 != fields.size()) {
+        stream_ << ", ";
+      }
+    }
+    stream_ << ')';
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const VarNode* op) final {
+    Var var = GetRef<Var>(op);
+    // This is an unbounded var.
+    TextValue val = AllocVarName(var);
+    this->PrintIndent();
+    stream_ << "free_var ";
+    this->PrintVarDecl(var, stream_);
+    this->PrintEndInst("\n");
+    return val;
+  }
+
+  TextValue VisitExpr_(const GlobalVarNode* op) final {
+    return TextValue('@' + op->name_hint);
+  }
+
+  TextValue VisitExpr_(const FunctionNode* op) final {
+    TextValue id = AllocTempVar();
+    std::ostringstream os;
+    os << id << " = function";
+    this->PrintFuncInternal(os.str(), GetRef<Function>(op));
+    this->PrintEndInst("\n");
+    return id;
+  }
+
+  TextValue VisitExpr_(const CallNode* op) final {
+    // TODO(tqchen, marisha): support generic call
```

Comment:
marisa, or M.K.

Reply:
Thanks, also re the old doc.h I understand that the old doc.h is more powerful as it brings reformating, but it is a feature that we might not need given the extra complexity